File: services/identity/helpers/error_helper.go
package helpers

import (
	"fmt"

	"github.com/gin-gonic/gin"
)

func printError(errorMessage string, path string, status int, c *gin.Context) {
	fmt.Printf("\nERROR: %d\n", status)
	fmt.Printf("	Path: @%s\n", path)
	fmt.Printf("	Message: %s\n", errorMessage)
	fmt.Printf("	Origin: %s", c.Request.URL)
}

func NetworkError(c *gin.Context, errorMessage string, path string) {
	printError(errorMessage, path, 500, c)
	c.JSON(500, gin.H{
		"success": false,
		"message": "Oops! Looks like an error occured on our end.",
		"code":    "INVALID_SERVER_ERROR",
	})
}

func AuthError(c *gin.Context, errorMessage string, path string) {
	printError(errorMessage, path, 401, c)
	c.JSON(401, gin.H{
		"message": "Authentication required. Please log in or provide a valid token.",
		"code":    "INVALID_AUTHORIZATION_ERROR",
	})
}

func BadRequestBodyError(c *gin.Context, errorMessage string, path string) {
	printError(errorMessage, path, 400, c)
	c.JSON(400, gin.H{
		"success": false,
		"message": "Invalid request body. Please check the data you provided.",
		"code":    "INVALID_REQUEST_BODY",
	})
}

func BadRequestParamsError(c *gin.Context, errorMessage string, path string) {
	printError(errorMessage, path, 400, c)
	c.JSON(400, gin.H{
		"success": false,
		"message": "Invalid URL parameters. Please verify the request path.",
		"code":    "INVALID_REQUEST_PARAMS",
	})
}

func BadRequestQueryError(c *gin.Context, errorMessage string, path string) {
	printError(errorMessage, path, 500, c)
	c.JSON(400, gin.H{
		"success": false,
		"message": "Invalid query parameters. Please check your query values.",
		"code":    "INVALID_REQUEST_QUERY",
	})
}

func ForbiddenError(c *gin.Context, errorMessage string, path string) {
	printError(errorMessage, path, 403, c)
	c.JSON(403, gin.H{
		"success": false,
		"message": "Access denied. You do not have permission to perform this action.",
		"code":    "FORBIDDEN",
	})
}

func NotFoundError(c *gin.Context, errorMessage string, path string) {
	printError(errorMessage, path, 404, c)
	c.JSON(404, gin.H{
		"success": false,
		"message": "The requested resource could not be found.",
		"code":    "NOT_FOUND",
	})
}

File: services/content/src/helpers/errorHelper.ts
interface ErrorDetails {
    (message: string,
    status: number,
    filename: string,
    json: Record<string, any>): void | HttpError
}

interface HttpErrorDetails {
    message: string,
    status: number,
    filename: string,
    json: Record<string, any>
}

class HttpError extends Error implements HttpErrorDetails{
  status: number
  filename: string
  json: Record<string, any>

  constructor(message: string, status: number, filename: string, json: Record<any, string>) {
    super(message)
    this.status = status
    this.json = json
    this.filename = filename
    Object.setPrototypeOf(this, HttpError.prototype)
  }
}

const throwError: ErrorDetails = (message, status, filename, json) => {
  throw new HttpError(message, status, filename, json)
}

export const returnError: ErrorDetails = (message, status, filename, json) => {
  return new HttpError(message, status, filename, json)
}


export const errorHelpers = {
  networkError: (errorMessage: string, filename: string) => 
    throwError(errorMessage, 500, filename, {
      message: "Oops! Looks like an error occured on our end.",
      code: "INVALID_SERVER_ERROR"
  }),

  authError: (errorMessage: string, filename: string) => 
    throwError(errorMessage, 401, filename, {
      message: "Authentication required. Please log in or provide a valid token.",
      code: "INVALID_AUTHORIZATION_ERROR"
  }),
  genericBadRequestError: (errorMessage: string, filename: string, validationErrors: any[]) =>
    throwError(errorMessage, 400, filename, {
      message: "Invalid request. Please check the data you provided.",
      code: "INVALID_REQUEST",
      validationErrors: validationErrors
    }),
  badRequestBodyError: (errorMessage: string, filename: string, validationErrors?: any[]) => 
    throwError(errorMessage, 400, filename, {
      message: "Invalid request body. Please check the data you provided.",
      code: "INVALID_REQUEST_BODY",
      validationErrors: validationErrors ?? null
    }),

  badRequestParamsError: (errorMessage: string, filename: string, validationErrors?: any[]) => 
    throwError(errorMessage, 400, filename, {
      message: "Invalid URL parameters. Please verify the request path.",
      code: "INVALID_REQUEST_PARAMS",
      validationErrors: validationErrors ?? null
  }),

  badRequestQueryError: (errorMessage: string, filename: string, validationErrors? : any[]) => 
    throwError(errorMessage, 400, filename, {
      message: "Invalid query parameters. Please check your query values.",
      code: "INVALID_REQUEST_QUERY",
      validationErrors: validationErrors ?? []
  }),

  forbiddenError: (errorMessage: string, filename: string) => 
    throwError(errorMessage, 403, filename, {
      message: "Access denied. You do not have permission to perform this action.",
      code: "FORBIDDEN"
  }),
  notFoundError: (errorMessage: string, filename: string) => 
    throwError(errorMessage, 404, filename, {
      message: "The requested resource could not be found.",
      code: "NOT_FOUND"
  })
};
export default throwError